src/java.base/share/classes/java/security/Security.java

```diff
@@ -592,40 +593,28 @@ public static Provider[] getProviders(Map<String,String> filter) {
         // Then only return those providers who satisfy the selection criteria.
         Provider[] allProviders = Security.getProviders();
         Set<String> keySet = filter.keySet();
-        LinkedHashSet<Provider> candidates = new LinkedHashSet<>(5);
 
         // Returns all installed providers
         // if the selection criteria is null.
         if ((keySet == null) || (allProviders == null)) {
```

Comment:
Both may be null as their javadoc didn't state explicitly what is returned.
This is existing code and it returns null upon empty keySet which I keep the same behavior. 
Perhaps for consistency sake, it should also check for the empty keySet condition and return allProviders.

Reply:
Well, it depends on how you interpret an empty filter map. If an empty table is specified, I'd think it means no criteria is set and all providers can be returned as the comment above suggested.
As this is a corner case, I am ok with reverting the isEmpty() check to keep the same behavior as before. However, I don't get which part of spec states about returning null for this case?

Reply:
I see that null is returned instead of an empty Provider array. When an empty filter map is specified, I interpret it as no criteria is set and thus it should return all providers (which is often non-empty). The condition on line 599 is OR, but the line on 600 returns the "allProvider" value which may or may not be null. Anyway, I start to get the feeling that we may not be talking about the same stuff...

Reply:
So, is it about empty all Providers or empty filter map? I only added the check on latter. Are you suggesting to add a check for the former?

Reply:
Then I will re-write the checks. The original code does not ensure null is returned when allProviders is empty.